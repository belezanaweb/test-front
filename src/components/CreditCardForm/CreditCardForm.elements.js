import styled from 'styled-components'

export const FormTitle = styled.span`
  margin: 2rem 1rem 0.5rem;
  color: #999;
  font-weight: bold;
  display: block;
`

export const FormContainer = styled.form`
  padding: 2rem;
  background-color: #fff;
  border-radius: 3px;
  box-shadow: 1px 1px 5px rgba(0, 0, 29, 0.22);

  display: flex;
  flex-direction: column;

  & div + div {
    margin-top: 2rem;
  }

  @media (max-width: 720px) {
    padding: 1rem;

    & div + div {
      margin-top: 1rem;
    }
  }
`

export const FormInputContainer = styled.div`
  display: flex;
  flex-direction: column;
`

export const FormInputLabel = styled.label`
  font-size: 1rem;
  font-weight: bold;
  line-height: 0.875rem;
  color: #ccc;
  margin-bottom: 0.5rem;

  @media (max-width: 720px) {
    font-size: 0.75rem;
  }
`

export const FormInput = styled.input`
  width: 100%;
  font-size: 1rem;
  padding: 0.5rem;
  height: 2.875rem;
  border-radius: 3px;
  border: 1px solid #e7e7e7;
  box-shadow: inset 0 1px 2px 0 rgba(0, 0, 0, 0.2);

  &::placeholder {
    color: #e0e7ee;
  }
`

export const FormInputContainerSpecial = styled.div`
  width: 100%;
  display: flex;

  & div + div {
    margin-left: 2rem;
  }

  & div:last-child {
    margin-top: 0;
  }
`
