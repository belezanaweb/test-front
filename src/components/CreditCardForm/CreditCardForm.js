import React from 'react'
import Input from '../Input/Input'
import { createUseStyles } from 'react-jss'

const useStyles = createUseStyles({
  creditCardForm: {
    width: '100%',
    padding: '12px 10px 0',
    margin: '0 auto',
    boxSizing: 'border-box',
    display: 'flex',
    '& form': {
      width: '100%'
    }
  },
  doubleColumn: {
    display: 'flex',
    flexWrap: 'wrap',
    '& div:first-child': {
      width: '55%',
      marginRight: '5%'
    },
    '& div:last-child': {
      width: '40%'
    }
  }
})

const CreditCardForm = (props) => {
  const classes = useStyles()
  const { disableSubmit } = props

  return (
    <div className={classes.creditCardForm} data-testid="CreditCardForm">
      <form>
        <Input labelText="Número do Cartão" type="number" role="creditCard" />

        <Input labelText="Nome do Titular" type="text" role="holderName" />

        <div className={classes.doubleColumn}>
          <Input labelText="CVV" type="number" role="cvv" />
          <Input labelText="Validade (mês/ano)" type="text" role="expiryDate" />
        </div>
      </form>
    </div>
  )
}

export default CreditCardForm
