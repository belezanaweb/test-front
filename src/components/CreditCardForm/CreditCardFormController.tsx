import React from 'react';
import { useForm, FormContextValues } from "react-hook-form";

interface CreditCardFormContextProps extends Pick<FormContextValues, 'control' | 'register' | 'errors'> {}

export const CreditCardFormContext = React.createContext<CreditCardFormContextProps | null>(null);

interface CreditCardFormControllerProps {
  children: (...args: any) => React.ReactNode;
  onSubmit: (...args: any) => void;
}

const CreditCardFormController: React.FC<CreditCardFormControllerProps> = ({ children, onSubmit }) => {
  const { control, errors, handleSubmit, register } = useForm();

  return (
    <CreditCardFormContext.Provider value={{ control, register, errors }}>
      <form onSubmit={handleSubmit(onSubmit)}>
        {children({ errors })}
      </form>
    </CreditCardFormContext.Provider>
  );
};

export default CreditCardFormController;
