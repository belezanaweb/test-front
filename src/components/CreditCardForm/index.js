import React, { useState, useCallback } from 'react'

import { useCreditCard } from '../../hooks/creditCard'

import Card from '../Card'
import Input from '../Input'

import { Form, Warper } from './styles'

const CreditCardForm = ({ formId, formSubmitted }) => {
  const { formData } = useCreditCard()

  const [error, setError] = useState({})

  const handleSubmit = (e) => {
    e.preventDefault()
    if (FormDataIsValid()) formSubmitted()
  }

  const FormDataIsValid = useCallback(() => {
    let errorObj = { valid: false }

    Object.keys(formData).forEach((data) => {
      let isValid

      switch (data) {
        case 'number':
          isValid = formData.number
            ?.replaceAll('.', '')
            .match(
              /^(?:(4[0-9]{12}(?:[0-9]{3})?)|(5[1-5][0-9]{14})|(6(?:011|5[0-9]{2})[0-9]{12})|(3[47][0-9]{13})|(3(?:0[0-5]|[68][0-9])[0-9]{11})|((?:2131|1800|35[0-9]{3})[0-9]{11}))$/g
            )
          break
        case 'name':
          isValid = formData.name?.match(/^[a-zA-Z]{2,30}$/g)
          break
        case 'expiry':
          isValid = formData.expiry?.match(/[0-9/]{7}/g)
          break
        case 'cvv':
          isValid = formData.cvv?.match(/\d{3}/g)
          break
        default:
          break
      }
      errorObj = { ...errorObj, [data]: !isValid, valid: isValid }
    })

    setError(errorObj)

    return errorObj.valid
  }, [formData])

  return (
    <Card>
      <Form id={formId} onSubmit={handleSubmit}>
        <Input
          name={'number'}
          onlyNumbers
          placeholder={'____.____.____.____'}
          maskPattern={'####.####.####.####'}
          maskDivider={'.'}
          label={'Número do cartão:'}
          hasError={error.number}
        />
        <Input
          name={'name'}
          placeholder={'Como no cartão'}
          label={'Nome do Titular:'}
          hasError={error.name}
        />
        <Warper>
          <Input
            name={'expiry'}
            onlyNumbers
            placeholder={'__/____'}
            maskPattern={'##/####'}
            maskDivider={'/'}
            label={'Validade (mês/ano):'}
            hasError={error.expiry}
          />
          <Input
            name={'cvv'}
            onlyNumbers
            placeholder={'___'}
            maskPattern={'###'}
            label={'CVV:'}
            hasError={error.cvv}
          />
        </Warper>
      </Form>
    </Card>
  )
}

export default CreditCardForm
