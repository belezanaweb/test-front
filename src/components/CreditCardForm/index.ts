export { default } from './CreditCardForm';
export { default as CreditCardFormController, CreditCardFormContext } from './CreditCardFormController';
