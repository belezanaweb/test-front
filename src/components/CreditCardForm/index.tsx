import React from 'react'

import { useCreditCard } from '../../hooks/creditCard'

import Card from '../Card'
import Input from '../Input'

import { Form, Warper } from './styles'

interface Props {
  formId: string
  formSubmitted: () => void
}

const CreditCardForm: React.FC<Props> = ({ formId, formSubmitted }) => {

  const { error } = useCreditCard()

  const handleSubmit = (e: React.FormEvent<HTMLFormElement> | undefined) => {
    e?.preventDefault()
    formSubmitted()
  }

  return (
    <Card>
      <Form id={formId} onSubmit={handleSubmit}>
        <Input
          name={'number'}
          onlyNumbers
          placeholder={'____.____.____.____'}
          maskPattern={'####.####.####.####'}
          maskDivider={'.'}
          label={'Número do cartão:'}
          errorLabel={'Número do cartão inválido'}
          hasError={error.number}
        />
        <Input
          name={'name'}
          placeholder={'Como no cartão'}
          label={'Nome do Titular:'}
          errorLabel={'Nome de titular inválido'}
          hasError={error.name}
        />
        <Warper>
          <Input
            name={'expiry'}
            onlyNumbers
            placeholder={'__/____'}
            maskPattern={'##/####'}
            maskDivider={'/'}
            label={'Validade (mês/ano):'}
            errorLabel={'Data inválida'}
            hasError={error.expiry}
          />
          <Input
            name={'cvv'}
            onlyNumbers
            placeholder={'___'}
            maskPattern={'###'}
            label={'CVV:'}
            errorLabel={'CVV inválido'}
            hasError={error.cvv}
          />
        </Warper>
      </Form>
    </Card>
  )
}

export default CreditCardForm
