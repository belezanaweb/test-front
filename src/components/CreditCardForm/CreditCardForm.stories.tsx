import React from 'react';
import { action } from '@storybook/addon-actions';

import theme, { Theming } from '../../theme';

import Box from '../Box';

import CreditCardForm from './CreditCardForm';
import CreditCardFormController from './CreditCardFormController';

const Provider = (storyFn: () => React.ReactNode) => (
  <Theming theme={theme}>
    {storyFn()}
  </Theming>
);

export default {
  component: CreditCardForm,
  decorators: [Provider],
  title: 'CreditCardForm',
};

export const Default = () => (
  <Box css={{ padding: '20px' }}>
    <CreditCardFormController onSubmit={action('submit')}>
      {() => (
        <>
          <CreditCardForm />
          <button type="submit">Submit</button>
        </>
      )}
    </CreditCardFormController>
  </Box>
);
