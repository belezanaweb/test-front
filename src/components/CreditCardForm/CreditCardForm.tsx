import React, { useContext } from 'react';
import { Controller } from 'react-hook-form';
import styled from 'styled-components';

import Box, { Grid } from '../Box';
import TextField from '../TextField';

import { CreditCardFormContext } from './CreditCardFormController';

const StyledForm = styled.div`
  flex: 1;
`;

const CreditCardForm = () => {
  const context = useContext(CreditCardFormContext);

  if (context === null) return null;

  const { control, errors } = context;

  function renderFieldNumber() {
    const mask = [/\d/, /\d/, /\d/, /\d/, '.', /\d/, /\d/, /\d/, /\d/, '.', /\d/, /\d/, /\d/, /\d/, '.', /\d/, /\d/, /\d/, /\d/];
    return (
      <Controller
        as={
          <TextField
            label="Número do Cartão:"
            mask={mask}
            error={errors.number && errors.number.message}
            placeholder="____.____.____.____"
          />
        }
        name="number"
        rules={{ required: "Numero do cartão obrigatório." }}
        control={control}
      />
    );
  }

  function renderFieldName() {
    return (
      <Controller
        as={
          <TextField
            label="Nome do Titular:"
            mask={false}
            error={errors.name && errors.name.message}
            placeholder="Como no cartão"
          />
        }
        name="name"
        rules={{ required: "Nome do titular obrigatório." }}
        control={control}
      />
    );
  }

  function renderFieldExpiringDate() {
    return (
      <Controller
        as={
          <TextField
            label="Validade (mês/ano):"
            mask={[/\d/, /\d/, '/', /\d/, /\d/, /\d/, /\d/]}
            error={errors.expiringDate && errors.expiringDate.message}
            placeholder="__/____"
          />
        }
        name="expiringDate"
        rules={{ required: "Validade do cartão obrigatório." }}
        control={control}
      />
    );
  }

  function renderFieldCodeVerification() {
    return (
      <Controller
        as={
          <TextField
            error={errors.code && errors.code.message}
            label="CVV:"
            mask={[/\d/, /\d/, /\d/]}
            placeholder="___"
          />
        }
        name="code"
        rules={{ required: "Codígo de verificação obrigatório." }}
        control={control}
      />
    );
  }

  return (
    <StyledForm>
      <Grid template="1fr" gap="25px" css={{ marginBottom: '25px' }}>
        <Box>
          {renderFieldNumber()}
        </Box>

        <Box>
          {renderFieldName()}
        </Box>
      </Grid>

      <Grid template="1.2fr 1fr" css={{ marginBottom: '15px' }}>
        <Box>
          {renderFieldExpiringDate()}
        </Box>
        <Box>
          {renderFieldCodeVerification()}
        </Box>
      </Grid>
    </StyledForm>
  );
};

export default CreditCardForm;
