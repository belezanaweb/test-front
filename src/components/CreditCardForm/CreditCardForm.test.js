import React from 'react'
import renderer from 'react-test-renderer'
import CreditCardForm from './CreditCardForm'

it('renders correctly', () => {
  const tree = renderer.create(<CreditCardForm></CreditCardForm>).toJSON()
  expect(tree).toMatchSnapshot()
})
