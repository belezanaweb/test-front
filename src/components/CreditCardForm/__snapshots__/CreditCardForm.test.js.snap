// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="creditCardForm-0-2-1"
  data-testid="CreditCardForm"
>
  <form>
    <div
      className="inputWrapper-0-2-4"
    >
      <label>
        Número do Cartão
      </label>
      <input
        className="errorInput-0-2-3"
        onChange={[Function]}
        type="number"
      />
      <p
        className="errorMessage-0-2-5"
      >
        
      </p>
    </div>
    <div
      className="inputWrapper-0-2-4"
    >
      <label>
        Nome do Titular
      </label>
      <input
        className="errorInput-0-2-3"
        onChange={[Function]}
        type="text"
      />
      <p
        className="errorMessage-0-2-5"
      >
        
      </p>
    </div>
    <div
      className="doubleColumn-0-2-2"
    >
      <div
        className="inputWrapper-0-2-4"
      >
        <label>
          CVV
        </label>
        <input
          className="errorInput-0-2-3"
          onChange={[Function]}
          type="number"
        />
        <p
          className="errorMessage-0-2-5"
        >
          
        </p>
      </div>
      <div
        className="inputWrapper-0-2-4"
      >
        <label>
          Validade (mês/ano)
        </label>
        <input
          className="errorInput-0-2-3"
          onChange={[Function]}
          type="text"
        />
        <p
          className="errorMessage-0-2-5"
        >
          
        </p>
      </div>
    </div>
  </form>
</div>
`;
